File: python/src/image_processor_pkg/image_processor_pkg/subscriber.py
#!/usr/bin/python3
import rclpy
from rclpy.node import Node
from image_processor_pkg.msg import ObjectLocation

# Perfil para QoS preconfigurado, tiene:
#   History: Keep last,
#   Depth: 5,
#   Reliability: Best effort,
#   Durability: Volatile,
#   Deadline: Default,
#   Lifespan: Default,
#   Liveliness: System default,
#   Liveliness lease duration: default,
#   avoid ros namespace conventions: false
# Informacion sacada de: https://docs.ros2.org/latest/api/rclcpp/classrclcpp_1_1SensorDataQoS.html
from rclpy.qos import qos_profile_sensor_data
import os
import csv


class PositionReceiver(Node):
    def __init__(self):
        super().__init__("position_receiver")

        # Ruta base donde querermos guardar los csv
        self.csv_base_path = "/ros2_ws/src/image_processor_pkg/mediciones_python_"

        #Diccionario para gestionar los escritores de csv
        self.recursos = {}

        self.preparar_csv()

        # 2. Suscripción
        self.subscription = self.create_subscription(
            ObjectLocation,
            "/object_position",
            self.topic_callback,
            qos_profile_sensor_data,
        )

        self.get_logger().info("Node PositionReceiver Ready")

    def obtener_recurso(self,nodo_id):
        if nodo_id not in self.recursos:
            file_path = f"{self.csv_base_path}{nodo_id}.csv"
            
            # Comprobación de cabecera
            necesita_cabecera = not os.path.exists(file_path) or os.stat(file_path).st_size == 0
            
            f = open(file_path, mode="a", newline="")
            writer = csv.writer(f)
            
            if necesita_cabecera:
                writer.writerow([
                    "timestamp_ns", "color", "cpu_proc_ms", 
                    "network_lat_ms", "total_lat_ms"
                ])
                # Para forzar a que haga la escritura a disco para asegurar tener cabecera
                f.flush()

            # Guardamos la pareja (archivo, escritor) en el diccionario
            self.recursos[nodo_id] = (f, writer)

        return self.recursos[nodo_id]

    def topic_callback(self, msg):
        # Tiempo en el que se recibe el mensaje
        tiempo_recibido = self.get_clock().now()
        # Tiempo en que el mensaje salió del emisor
        tiempo_envio = rclpy.time.Time.from_msg(msg.stamp)

        # 1. Latencia de red pura (tiempo de viaje)
        latencia_red_ns = tiempo_recibido - tiempo_envio
        latencia_red_ms = latencia_red_ns.nanoseconds / 1e6

        # 2. Tiempo de procesamiento en la otra RPi (viene en el mensaje)
        cpu_ms = msg.proc_time

        # 3. Latencia Total (desde que la cámara captó el frame hasta que llegó aquí)
        total_ms = cpu_ms + latencia_red_ms

        # --- GUARDAR DATOS ---
        _, writer = self.obtener_recurso(msg.node_id)

        writer.writerow([
            tiempo_recibido.nanoseconds,
            msg.color,
            f"{cpu_ms:.4f}",
            f"{latencia_red_ms:.4f}",
            f"{total_ms:.4f}",
        ])

        # --- LOGS POR CONSOLA ---
        self.get_logger().info(
            f"RECIBIDO -> Color: {msg.color} | CPU: {cpu_ms:.2f}ms | RED: {latencia_red_ms:.2f}ms"
        )

    def destroy_node(self):
        """ Cerramos los ficheros abiertos, iterando sobre el diccionario """
        for robot_id, (f_obj, _) in self.recursos.items():
            f_obj.close()
            self.get_logger().info(f"Archivo cerrado: ID {robot_id}")
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)

    position_receiver = PositionReceiver()

    try:
        # Mantiene el nodo abierto procesando los mensajes entrantes
        rclpy.spin(position_receiver)
    except KeyboardInterrupt:
        pass
    finally:
        position_receiver.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: c++/src/image_processor_pkg/src/subscriber.cpp
#include "rclcpp/rclcpp.hpp"
#include "image_processor_pkg/msg/object_location.hpp"
#include <fstream>
#include <iostream>
#include <map>
#include <memory>
#include <string>
#include <filesystem> // Para verificar si el archivo existe y su tamaño

class PositionReceiver : public rclcpp::Node {
public:
  PositionReceiver() : Node("position_receiver") {
    // Ruta base para los archivos CSV
    csv_base_path = "/ros2_ws/src/image_processor_pkg/mediciones_cpp_";

    // Nos subscribimos al topic que publica la informacion
    sub = this->create_subscription<image_processor_pkg::msg::ObjectLocation>(
        "/object_position", rclcpp::SensorDataQoS(),
        std::bind(&PositionReceiver::topic_callback, this, std::placeholders::_1));

    RCLCPP_INFO(this->get_logger(), "Nodo PositionReceiver C++ listo.");
  }

  // Destructor para cerrar los archivos correctamente al terminar
  ~PositionReceiver() {
    for (auto const& [id, file_ptr] : recursos) {
      if (file_ptr->is_open()) {
        file_ptr->close();
      }
    }
  }

private:
  // Función para obtener o crear el recurso de escritura
  std::ofstream& obtener_recurso(const std::string& node_id) {

    // Comprobamos si ya esta abierto el fichero. El recurso esta guardado en el diccinario
    if (recursos.find(node_id) == recursos.end()) {
      std::string file_path = csv_base_path + node_id + ".csv";
      
      // Comprobar si necesita cabecera (si no existe o está vacío)
      bool necesita_cabecera = !std::filesystem::exists(file_path) || std::filesystem::file_size(file_path) == 0;
      
      // Creamos el puntero al archivo en modo append
      auto f = std::make_unique<std::ofstream>(file_path, std::ios::app);
      
      if (necesita_cabecera) {
          // Añadimos la cabecera al fichero
          *f << "timestamp_ns,color,pos_x,pos_y,cpu_proc_ms,network_lat_ms,total_lat_ms\n";
          // Forzamos la escritura de la cabecera para evitar errores
          f->flush();
      }
      
      recursos[node_id] = std::move(f);
    }
    
    // Devolvemos el puntero al recuros donde guardar los datos
    return *recursos[node_id];
  }

  void topic_callback(const image_processor_pkg::msg::ObjectLocation::SharedPtr msg) {

    // Tiempo actual
    auto now = this->now();
    // Tiempo de envío desde el mensaje
    auto stamp = rclcpp::Time(msg->stamp);

    // 1. Latencia de red pura
    double latencia_red_ms = (now - stamp).nanoseconds() / 1e6;
    
    // 2. Tiempo de procesamiento (del mensaje)
    double cpu_ms = msg->proc_time;
    
    // 3. Latencia Total
    double total_ms = cpu_ms + latencia_red_ms;

    // --- GUARDAR DATOS ---
    std::ofstream& f = obtener_recurso(msg->node_id);

    f << std::fixed << std::setprecision(4);
    f << now.nanoseconds() << "," 
      << msg->color << "," 
      << msg->x << ","        
      << msg->y << ","        
      << cpu_ms << ","
      << latencia_red_ms << "," 
      << total_ms << "\n";

    // --- LOGS POR CONSOLA ---
    RCLCPP_INFO(this->get_logger(),
                "RECIBIDO -> Color: %s | CPU: %.2fms | RED: %.2fms",
                msg->color.c_str(), cpu_ms, latencia_red_ms);
  }

  std::string csv_base_path;
  // Diccionario (map) para gestionar los escritores de CSV
  std::map<std::string, std::unique_ptr<std::ofstream>> recursos;
  rclcpp::Subscription<image_processor_pkg::msg::ObjectLocation>::SharedPtr sub;
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<PositionReceiver>());
  rclcpp::shutdown();
  return 0;
}